File: 프로그래머스/lv2/42586. 기능개발/기능개발.py
# 아이디어 : 1. 두 리스트를 큐(덱)로 변경
#           2. 
# 알고리즘 : 
# 자료구조 : queue(deque)

from collections import deque

def solution(progresses, speeds):
    answer = []
    progresses = deque(progresses)
    speeds = deque(speeds)
    day = 1 # 현재 날짜
    cnt = 0 # 배포 기능 개수
    
    while progresses :
        # 배포가 가능하다면
        while len(progresses) > 0 : 
            if progresses[0] + day * speeds[0] >= 100 :
                progresses.popleft()
                speeds.popleft()
                cnt += 1
            else :
                break
        
        # 배포할 기능이 있다면
        if cnt > 0 :
            answer.append(cnt)
            cnt = 0
        day += 1
    
    return answer

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();
        
        Queue<Data> queue = new LinkedList<>();
        
        for (int i = 0; i < progresses.length; i++) {
            queue.offer(new Data(progresses[i], speeds[i]));
        }
        
        int day = 1;
        int cnt = 0;
        
        while(!queue.isEmpty()) {
            Data data = queue.peek();
            int p = data.getProgress();
            int s = data.getSpeed();
            
            if ((p + day * s) >= 100) {
                queue.poll();
                cnt += 1;
            } else {
                if (cnt > 0) {
                    answer.add(cnt);
                    cnt = 0;    
                }
               day += 1;     
            }
        } 
        if (cnt > 0) 
            answer.add(cnt);
        
        return answer.stream().mapToInt(x -> x).toArray();
    }
}

class Data {
    private int progress;
    private int speed;
    
    public Data(int progress, int speed) {
        this.progress = progress;
        this.speed = speed;
    }
    
    int getProgress() {
        return progress;
    }
    
    int getSpeed() {
        return speed;
    }
}